pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -580,45 +684,39 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListenerForPortOrUDS(no
 		switch pluginParams.ListenerProtocol {
 		case plugin.ListenerProtocolHTTP:
 			filterChainMatch = chain.FilterChainMatch
+			if filterChainMatch != nil && len(filterChainMatch.ApplicationProtocols) > 0 {
+				// This is the filter chain used by permissive mTLS. Append mtlsHTTPALPNs as the client side will
+				// override the ALPN with mtlsHTTPALPNs.
+				// TODO: This should move to authN code instead of us appending additional ALPNs here.
+				filterChainMatch.ApplicationProtocols = append(filterChainMatch.ApplicationProtocols, mtlsHTTPALPNs...)
+			}
 			httpOpts = configgen.buildSidecarInboundHTTPListenerOptsForPortOrUDS(node, pluginParams)
 
 		case plugin.ListenerProtocolTCP:
 			filterChainMatch = chain.FilterChainMatch
 			tcpNetworkFilters = buildInboundNetworkFilters(pluginParams.Env, pluginParams.Node, pluginParams.ServiceInstance)
 
 		case plugin.ListenerProtocolAuto:
-			// TODO(crazyxy) avoid bypassing authN using TCP
-			// Build filter chain options for listener configured with protocol sniffing
-			// Double the number of filter chains. Half of filter chains are used as http filter chain and half of them are used as tcp proxy
-			// id in [0, len(allChains)/2) are configured as http filter chain, [(len(allChains)/2, len(allChains)) are configured as tcp proxy
-			// If mTLS is enabled, there are four filter chains. The filter chain match should be
-			//  FCM 1: ALPN [istio, http/1.0, http/1.1, h2] Transport protocol: tls
-			//  FCM 2: ALPN [http/1.0, http/1.1, h2] Transport protocol: N/A
-			//  FCM 3: ALPN [istio] Transport protocol: N/A
-			//  FCM 4: ALPN [] Transport protocol: N/A
-			// If mTLS is disabled, there are two filter chains. The filter chaim match should be
-			//  FCM 1: ALPN [http/1.0, http/1.1, h2]
-			//  FCM 2: ALPN []
-			if id < len(allChains)/2 {
-				httpOpts = configgen.buildSidecarInboundHTTPListenerOptsForPortOrUDS(node, pluginParams)
-
-				fcm := listener.FilterChainMatch{}
-				if chain.FilterChainMatch != nil {
-					fcm = *chain.FilterChainMatch
-				}
+			// Make sure id is not out of boundary of filterChainMatchOption
+			if filterChainMatchOption == nil || len(filterChainMatchOption) <= id {
+				continue
+			}
 
-				fcm.ApplicationProtocols = append(fcm.ApplicationProtocols, applicationProtocols...)
-				filterChainMatch = &fcm
+			// TODO(yxue) avoid bypassing authN using TCP
```

Comment:
https://github.com/istio/istio/issues/16214 It similar to the issue discussed in the email. Clients can fake the request to bypass the HTTP request along with the authN applied to the HTTP filter chain,

Reply:
Could you clarify a little how is authN being bypassed using TCP? If this feature is going to be enabled by default, this sounds a potential vulnerability and need to be resolved before this feature is enabled by default in 1.4.

cc @diemtvu @incfly

Reply:
There are multiple ways to hack the server. 
    1. You can send TCP traffic to bypass the HTTP autnN
    2. You can fake a client to send request slowly until the protocol detection timeout, it will fallback to the default filter chain which will bypass the security check.

Reply:
For case 1, shouldn't the TCP traffic be rejected because it's not matched with the HTTP filter chain?
For case 2, I would image the similar thing.

Reply:
If we reject TCP traffic if it's not matched with the HTTP filter chain, why not always creating filter chain? The problem is that we don't know the traffic from the client.

Reply:
just catching up from the conversation with yangmin. The conclusion is that authn is service based so we don't have the duty to protect requests on ports not declared in the service definition.

we'll need to solve this for workload based jwt policy @diemtvu i think.

Reply:
should the comments in the code to reflect that?

Reply:
> The conclusion is that authn is service based so we don't have the duty to protect requests on ports not declared in the service definition.

@PiotrSikora @rshriram @howardjohn thoughts?

Reply:
We do, however, need to make sure the ongoing new JWT policy implementation covers this case. Because the new JWT policy will be workload based, the undeclared http port traffic should be in scope as well.

(if the intended behavior is to use authorization to reject the request, we should not have false bypass, but we can falsely reject the request due to this timeout and fall back to the tcp filter chain, IIUC)

@diemtvu fyi

Reply:
But we wouldn't reject the request if we fallback to TCP filter chain, since it doesn't have authN filters, correct?

Reply:
hmm... that seems true. very hard to deal with workload based jwt authentication when we're unable to reliably determine it's http traffic or not by sniffing.

unless we make port a requirement as for the new jwt policy...

Reply:
Well, yeah, that's my point, if authN / JWT policy is explicitly configured for a service, then that implies that the service is handling HTTP traffic and we don't need to guess the type of traffic anymore, i.e. we should only configure HTTP filter chain, without sniffing or fallback to TCP.

(I thought that I mentioned it here before, but it might have been in one of the email threads).

Reply:
I think the problem is after JWT policy is updated to workload-based, how do we prevent bypassing it using the passthrough TCP filter chain. In that case, the user expectation for the workload-based JWT policy is that it should enforce all traffic to the workload, even including the traffic captured by the passthrough TCP filter chain, and we cannot insert the JWT filter to the TCP filter chain in this case.

Reply:
AuthN/JWT policy can be set at service (and in future, workload) level, though it is expected to be applied on HTTP ports only. You will have trouble if the workload actually has TCP ports.

On top of that, authZ filter are also built differently between TCP and HTTP, so we may also apply the access control incorrectly if traffic is passed through the wrong channel.

Reply:
Can we move the discussion to this issue? https://github.com/istio/istio/issues/16214

The authN bypass issue is *not* introduced by this PR. authN and alpn filter are two fixes for inbound protocol sniffing, this PR is separate from the authN.